cpu/rpx0xx: port RIOT to the Raspberry Pi RP2040 MCU

maribu opened the pull request:
### Contribution description

This provides:

- Support for the RP2040 MCU
     - Only `periph_gpio`, `periph_gpio_int`, and `periph_uart` are provided in this PR. Other than that no timers, no SPI, no nothing is provided.
     - Work on additional drivers can more easily be parallelized once basic support is upstream, which this PR tries to do.
- Support for the Raspberry Pi Pico board (`rpi-pico`), the reference board for the RP2040 MCU, is provided

### Acknowledgements

This code was developed as part of a software project in a combined effort of students and mentors. Thanks to @nickw96, @MaestroOnICe, and @Franz2000.

Most of the low level code and especially the boot sequence and clock configuration was developed by @fabian18.

### Testing procedure

The following test applications will be useful:

- `examples/default`
    - stdio via UART should work
    - Via SAUL the on-board LED should be controllable
- `tests/periph_gpio`
    - The GPIO pins should work as input (floating, with pull up, with pull down) and output and IRQs should be supported on rising, falling, and both edges

It would be useful to test all of the flashing options:

- Using OpenOCD and `pico-debug` (currently the default option)
     - Beware: First "flash" the gimmechace variant from here https://github.com/majbthrd/pico-debug/releases via the UF2 bootloader into RAM. (Needs to be re-"flashed" after each cold boot, or RAM corruption.)
- Using J-Link
- Using the UF2 bootloader

### Issues/PRs references

~None~ closes #15822

benpicco commented on cpu/rp2040/periph/gpio.c:
```diff
@@ -0,0 +1,212 @@
+    volatile uint32_t *irq_status_regs = &IO_BANK0->PROC0_INTS0.reg;
+    volatile uint32_t *irq_ack_regs = &IO_BANK0->INTR0.reg;
+
+    for (unsigned i = 0; i < (GPIO_PIN_NUMOF + 7) / 8; i++, offset += 8) {
```
This needs some explaining - are there multiple GPIO banks after all?

maribu replied:
I added some comments now.

benpicco commented on cpu/rp2040/periph/gpio.c:
```diff
@@ -0,0 +1,212 @@
+    for (unsigned i = 0; i < (GPIO_PIN_NUMOF + 7) / 8; i++, offset += 8) {
+        unsigned status = irq_status_regs[i];
+        irq_ack_regs[i] = status;
+        for (unsigned pin = 0; pin < 8; pin++) {
+            if (status & (0xFLU << (pin << 2))) {
```
```suggestion
        uint8_t pin = 0;
        while (status) {
            status = bitarithm_test_and_clear(status, &pin);
            pin >>= 1;
```

should save us some iterations if only a single bit is set.

maribu replied:
There could actually be two bits set. I added some comments to explain.

benpicco commented on cpu/rp2040/periph/gpio.c:
```diff
@@ -0,0 +1,212 @@
+static void _irq_enable(unsigned pin, unsigned flank)
+{
+    volatile uint32_t *int_enable_regs = &IO_BANK0->PROC0_INTE0.reg;
+    unsigned shift_amount = (pin & 0x7) << 2;
+    unsigned idx = pin >> 3;
```
this could need a comment

maribu replied:
Added one

benpicco commented on cpu/rp2040/periph/uart.c:
```diff
@@ -0,0 +1,211 @@
+    io_config_reg_tx->function_select = FUNCTION_SELECT_UART;
+    if (ctx[uart].rx_cb) {
+        gpio_init(uart_config[uart].rx_pin, GPIO_IN_PU);
+        volatile gpio_io_ctrl_t *io_config_reg_rx = gpio_io_register(uart_config[uart].rx_pin);
+        io_config_reg_rx->function_select = FUNCTION_SELECT_UART;
```
Maybe we should rather have a `gpio_init_mux(uart_config[uart].rx_pin, FUNCTION_SELECT_UART)` so drivers don't have to call `gpio_io_register()` directly?

maribu replied:
Oha, that is actually a potentially corrupting write, as there is no guarantee that the compiler will use a 32 bit read modify write sequence for the access. I will reconsider the API altogether, as this looks very much like a valid use of the API.

maribu replied:
And there already is `gpio_set_function_select()` to safely do this.

benpicco commented on cpu/rp2040/periph/uart.c:
```diff
@@ -0,0 +1,211 @@
+                                        | UART0_UARTCR_TXE_Msk);
+    uart_init_pins(uart);
+
+    _irq_enable(uart);
```
`rx_cb` can be NULL, so we should only ~~enable the IRQ and~~ set `UART0_UARTCR_RXE_Msk` if it isn't

maribu replied:
Done

benpicco commented on cpu/rp2040/pll.c:
```diff
@@ -0,0 +1,131 @@
+ */
+static void _pll_stop(PLL_SYS_Type *pll)
+{
+    uint32_t reg = (1U << PLL_SYS_PWR_VCOPD_Pos) |
+                   (1U << PLL_SYS_PWR_POSTDIVPD_Pos) |
+                   (1U << PLL_SYS_PWR_DSMPD_Pos) |
+                   (1U << PLL_SYS_PWR_PD_Pos);
```
IMHO

```suggestion
    uint32_t reg = (1U << PLL_SYS_PWR_VCOPD_Pos)
                 | (1U << PLL_SYS_PWR_POSTDIVPD_Pos)
                 | (1U << PLL_SYS_PWR_DSMPD_Pos)
                 | (1U << PLL_SYS_PWR_PD_Pos);
```


looks nicer

benpicco commented on cpu/rp2040/system_RP2040.c:
```diff
@@ -0,0 +1,53 @@
+/*---------------------------------------------------------------------------
+  System Core Clock function
+ *---------------------------------------------------------------------------*/
+void SystemCoreClockUpdate (void)
+{
+    /* use frequency counter to check the clock frequency? */
+    SystemCoreClock = CLOCK_CORECLOCK;
+}
+
+/*---------------------------------------------------------------------------
+  System initialization function
+ *---------------------------------------------------------------------------*/
+void __attribute__((constructor)) SystemInit (void)
+{
+  SystemCoreClockUpdate();
+}
```
I think this is used nowhere in RIOT - or is this used by the vendor files?

maribu replied:
This is some CMSIS compatibility thing. I'm not sure, but since the vendor header files allow access to the system clock via the `SystemCoreClock` variable, it might makes sense to have that variable updated.

I hope that with LTO it will be detected that the variable is only written to, if no packages access the variable.

maribu replied:
On a second thought: If this is missing and used, the linker will tell what the issue is. I dropped the file entirely.

benpicco reviewed (commented):
Nice, looking good!
Some first comments:

maribu commented on boards/rpi-pico/include/board.h:
```diff
@@ -0,0 +1,52 @@
+ * @{
+ */
+#define LED0_PIN        GPIO_PIN(0, 25)
+#define LED0_ON         do {SIO->GPIO_OUT_SET.reg = 1UL << 25;} while(0)
```
An earlier version of the driver actually did this, but I decided that this is not worth the trouble it causes in the initialization code. After all, a bit-wise left shift is a simple operation, the inverse is (especially in absence of CLZ on Cortex M0+) not.

Also: The CPU cycle safed is not going to matter much, since the function call for `gpio_set()` etc. is going to be the main contributor to CPU time. If you use LTO to avoid this, then the compiler will also (in most case) be able to do the shift at compile time. So there isn't anything to be win.

IMO, if we want a fast GPIO API, we should go in the direction @gschorcht is heading with the new GPIO API.

maribu commented on boards/rpi-pico/Makefile.features:
```diff
@@ -0,0 +1,4 @@
+CPU = rp2040
```
Or maybe even rpx0, since RIOT is using only of the available cores anyway.

maribu replied:
renamed to rpx0xx

maribu commented on cpu/rp2040/boot_stage2/boot2_generic_03h.S:
```diff
@@ -0,0 +1,103 @@
+// ----------------------------------------------------------------------------
+// Second stage boot code
+// Copyright (c) 2019-2021 Raspberry Pi (Trading) Ltd.
+// SPDX-License-Identifier: BSD-3-Clause
```
Done

maribu commented on cpu/rp2040/include/periph_cpu.h:
```diff
@@ -0,0 +1,496 @@
+/**
+ * @brief   Clock for UART0 and UART1 peripherals
+ */
+#define CLOCK_CLKPERI       CLOCK_CORECLOCK
```
> Or is CLKPERI the name from the data sheet?

Yes, exactly. But I really like `CLOCK_PERIPH` better.

> Should it be possible to overwrite this if e.g. a board wants to clock the UART from a low power clock?

There are indeed quite a few options:

```C
/* ==========================================  CLOCKS CLK_PERI_CTRL AUXSRC [5..7]  =========================================== */
typedef enum {                                  /*!< CLOCKS_CLK_PERI_CTRL_AUXSRC                                               */
  CLOCKS_CLK_PERI_CTRL_AUXSRC_clk_sys  = 0,     /*!< clk_sys : clk_sys                                                         */
  CLOCKS_CLK_PERI_CTRL_AUXSRC_clksrc_pll_sys = 1,/*!< clksrc_pll_sys : clksrc_pll_sys                                          */
  CLOCKS_CLK_PERI_CTRL_AUXSRC_clksrc_pll_usb = 2,/*!< clksrc_pll_usb : clksrc_pll_usb                                          */
  CLOCKS_CLK_PERI_CTRL_AUXSRC_rosc_clksrc_ph = 3,/*!< rosc_clksrc_ph : rosc_clksrc_ph                                          */
  CLOCKS_CLK_PERI_CTRL_AUXSRC_xosc_clksrc = 4,  /*!< xosc_clksrc : xosc_clksrc                                                 */
  CLOCKS_CLK_PERI_CTRL_AUXSRC_clksrc_gpin0 = 5, /*!< clksrc_gpin0 : clksrc_gpin0                                               */
  CLOCKS_CLK_PERI_CTRL_AUXSRC_clksrc_gpin1 = 6, /*!< clksrc_gpin1 : clksrc_gpin1                                               */
} CLOCKS_CLK_PERI_CTRL_AUXSRC_Enum;
```

I changed the code to allow changing both source and clock speed with a define.

fabian18 commented on cpu/rp2040/cpu.c:
```diff
@@ -0,0 +1,89 @@
+    /* reset system PLL */
+    pll_reset_sys();
+    /* make the system PLL output a 125 MHz frequency from the 12 MHz XOSC */
+    pll_start_sys(MHZ(12), 1, 125, 6, 2);
```
PLL configuration values:
`12 MHz` input frequency from the `XTAL`,
divided by  `1`,
scaled with `125` by the `PLL`,
post-divided by `(6 * 2)` ,
makes `125` MHz.

fabian18 replied:
The settings are taken from the SDK.

maribu replied:
I moved the numbers to defines in `periph_cpu.h` and allowed to override those with defines. (E.g. up to 133 MHz is officially allowed, and some people clock their silicon with over 200 MHz, apparently to some success. So it might make sense to ease this.)

maribu commented:
@benpicco: May I do a first round of squashing?

benpicco commented:
sure, squash away!

maribu commented:
This is the diff from before the first squashing attempt and after the second attempt: https://github.com/RIOT-OS/RIOT/compare/86681da759ab8e2f57b3cb7fd2ba890f26148fc8..cafacf07845d3f1d0c6a3665943b75ebd55bbb21

(Sadly I incorrectly manually merged something on the first try.)

maribu commented:
The CI is now happy. Time for another round of squashing?

benpicco commented on boards/rpi-pico/doc.txt:
```diff
@@ -0,0 +1,138 @@
+Since option 2 requires no additional hardware, this is currently the default. However, you need to
+first "flash" the gimme-cache variant of [pico-debug](https://github.com/majbthrd/pico-debug)
+into RAM using the UF2 bootloader. This will result in the Raspberry Pi Pico showing up as
+CMSIS-DAP debugger. Afterwards run:
```
How is this done?
Or will the build system take care of that?

maribu replied:
The pico-debug .uf2 firmware needs to be copied into the virtual FAT flash drive. Upon unmounting that, the virtual flash drive will disappear and a CMSIS debugger will pop up. I could extend the documentation in this regard a bit, but I'd personally rather just keep the reference to pico-debug and let users read the upstream documentation.

This is not handled by the build system.

maribu replied:
OK, the upstream docu sucks in this regard. I'll add a sentence here.

benpicco commented on boards/rpi-pico/doc.txt:
```diff
@@ -0,0 +1,138 @@
+
+### Early state Implementation
+
+Currently no support for the following peripherals is implemented:
```
we should add the second CPU core to this list

maribu replied:
> What's the output of `tests/periph_gpio`'s `auto_test`?

```
$ make BOARD=rpi-pico flash test -C tests/periph_gpio
make: Entering directory '/home/maribu/Repos/software/RIOT/tests/periph_gpio'
There are unsatisfied feature requirements: periph_timer
/home/maribu/Repos/software/RIOT/tests/periph_gpio/../../Makefile.include:934: *** You can let the build continue on expected errors by setting CONTINUE_ON_EXPECTED_ERRORS=1 to the command line.  Stop.
make: Leaving directory '/home/maribu/Repos/software/RIOT/tests/periph_gpio'
```

:-/

maribu replied:
Done

benpicco commented on boards/rpi-pico/include/periph_conf.h:
```diff
@@ -0,0 +1,55 @@
+static const uart_conf_t uart_config[] = {
+    {
+        .dev = UART0,
+        .rx_pin = GPIO_PIN(0,1),
```
```suggestion
        .rx_pin = GPIO_PIN(0, 1),
```

benpicco commented on cpu/rpx0xx/Makefile.include:
```diff
@@ -0,0 +1,19 @@
+PICO_DEBUG      ?=
+ROM_LEN         ?= 2097152 # = 2 MiB used in the RPi Pico
```
since the MCU doesn't come with flash, shouldn't this always be set by the board?

maribu replied:
It is likely that most boards just use the reference flash chip, as the 2nd stage bootloader will then just work. Not every flash ship supports XIP in the same way, requiring other flash chips to use a different 2nd stage bootloader.

The board can still overwrite this value, if it indeed used a different value. IMO having the "reference flash chip" as default is sensible.

benpicco commented on cpu/rpx0xx/include/cpu_conf.h:
```diff
@@ -0,0 +1,47 @@
+ * @{
+ */
+#define CPU_DEFAULT_IRQ_PRIO            (1U)
+#define CPU_FLASH_BASE                  (0x10000000)
```
Should we re-use `ROM_START_ADDR` here?

benpicco commented on cpu/rpx0xx/periph/uart.c:
```diff
@@ -0,0 +1,211 @@
+#include "assert.h"
+
+static uart_isr_ctx_t ctx[UART_NUMOF];
+static mutex_t tx_lock = MUTEX_INIT_LOCKED;
```
not
```suggestion
static mutex_t tx_lock[UART_NUMOF];
```
?

maribu replied:
Oh, yes! Nice catch.

benpicco reviewed (commented):
looks good!
What's the output of `tests/periph_gpio`'s `auto_test`?

maribu commented on boards/rpi-pico/doc.txt:
```diff
@@ -0,0 +1,138 @@
+
+Currently (June 2021), only two methods for debugging via OpenOCD are supported:
+
+1. Using a bit-banging low-level adapter, e.g. via the GPIOs of a Raspberry Pi 4B
```
Theoretically yes, practically no. The issue is that multi-core CPUs need multi-drop support. And this needs to be implemented for every OpenOCD backend. E.g. the J-Link EDU mini can debug the RP2040 with the proprietary tool (so the hardware is capable), but fails to do so with OpenOCD due to missing multi-drop support in the JLink driver.

Currently, the situation is pretty dire regarding multi-drop support in OpenOCD. But this is likely going to change soon, with the RP2040 being relatively popular.

maribu commented:
> What's the output of `tests/periph_gpio`'s `auto_test`?

```
$ make BOARD=rpi-pico flash test -C tests/periph_gpio
make: Entering directory '/home/maribu/Repos/software/RIOT/tests/periph_gpio'
There are unsatisfied feature requirements: periph_timer
/home/maribu/Repos/software/RIOT/tests/periph_gpio/../../Makefile.include:934: *** You can let the build continue on expected errors by setting CONTINUE_ON_EXPECTED_ERRORS=1 to the command line.  Stop.
make: Leaving directory '/home/maribu/Repos/software/RIOT/tests/periph_gpio'
```

:-/

fabian18 commented on cpu/rpx0xx/periph/uart.c:
```diff
@@ -0,0 +1,215 @@
+        baud_fbrd = ((baud_rate_div & 0x7f) + 1) / 2;
+    }
+
+    io_reg_atomic_set(&(dev->UARTIBRD.reg), baud_ibrd & UART0_UARTIBRD_BAUD_DIVINT_Msk);
```
```suggestion
dev->UARTIBRD.reg = baud_ibrd;
```

fabian18 commented on cpu/rpx0xx/periph/uart.c:
```diff
@@ -0,0 +1,215 @@
+    }
+
+    io_reg_atomic_set(&(dev->UARTIBRD.reg), baud_ibrd & UART0_UARTIBRD_BAUD_DIVINT_Msk);
+    io_reg_atomic_set(&(dev->UARTFBRD.reg), baud_fbrd & UART0_UARTFBRD_BAUD_DIVFRAC_Msk);
```
```suggestion
dev->UARTFBRD.reg = baud_fbrd;
```

fabian18 commented on cpu/rpx0xx/periph/uart.c:
```diff
@@ -0,0 +1,215 @@
+
+    for (size_t i = 0; i < len; i++) {
+        dev->UARTDR.reg = data[i];
+        mutex_lock(&tx_locks[uart]);
```
The mutex lock leads to a deadlock for me if I try to `printf()` within an interrupt routine.
For example we could instead block if the UART Tx FIFO is full:
```C
while (dev->UARTFR.bit.TXFF) {}
```

maribu replied:
The FIFO is not used by the driver. It would have to busy wait until the char is send.

IMO we really should just disallow using UART during IRQ. We do so for every other periph, I don't know why UART is special in this regard.

maribu replied:
Fixed

benpicco commented:
> > What's the output of `tests/periph_gpio`'s `auto_test`?
> 
> ```
> $ make BOARD=rpi-pico flash test -C tests/periph_gpio
> make: Entering directory '/home/maribu/Repos/software/RIOT/tests/periph_gpio'
> There are unsatisfied feature requirements: periph_timer
> /home/maribu/Repos/software/RIOT/tests/periph_gpio/../../Makefile.include:934: *** You can let the build continue on expected errors by setting CONTINUE_ON_EXPECTED_ERRORS=1 to the command line.  Stop.
> make: Leaving directory '/home/maribu/Repos/software/RIOT/tests/periph_gpio'
> ```
> 
> :-/

you could try #14553 to use SysTick for xtimer

fabian18 commented:
> 

> > > What's the output of `tests/periph_gpio`'s `auto_test`?
> > 
> > 
> > ```
> > $ make BOARD=rpi-pico flash test -C tests/periph_gpio
> > make: Entering directory '/home/maribu/Repos/software/RIOT/tests/periph_gpio'
> > There are unsatisfied feature requirements: periph_timer
> > /home/maribu/Repos/software/RIOT/tests/periph_gpio/../../Makefile.include:934: *** You can let the build continue on expected errors by setting CONTINUE_ON_EXPECTED_ERRORS=1 to the command line.  Stop.
> > make: Leaving directory '/home/maribu/Repos/software/RIOT/tests/periph_gpio'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > :-/
> 
> you could try #14553 to use SysTick for xtimer

I took this as an occasion to upload #16627.

benpicco commented on cpu/rpx0xx/periph/uart.c:
```diff
@@ -0,0 +1,205 @@
+    io_reg_atomic_set(&dev->UARTCR.reg, enable_mask);
+    uart_init_pins(uart);
+
+    _irq_enable(uart);
```
Now you can move this to `if (rx_cb != NULL) {`

maribu commented on cpu/rpx0xx/periph/uart.c:
```diff
@@ -0,0 +1,201 @@
+    }
+}
+
+void uart_poweron(uart_t uart)
+{
+    uint32_t reset_bit_mask = (uart) ? RESETS_RESET_uart1_Msk : RESETS_RESET_uart0_Msk;
+    periph_reset(reset_bit_mask);
+    periph_reset_done(reset_bit_mask);
+}
+
+void uart_poweroff(uart_t uart)
+{
+    uart_deinit_pins(uart);
+    periph_reset((uart) ? RESETS_RESET_uart1_Msk : RESETS_RESET_uart0_Msk);
+}
```
Those are again wonderfully underspecified APIs. IMO, we should just drop `uart_poweron()`. Assume a UART is first initialized, then powered off, then power on again: Should it resume operation in the state before power off? Why should the driver need to sacrifice RAM to back up the configuration, when the user of the UART exactly knows all configuration details?

IMO, we should drop `uart_poweron()`. `uart_init()` powers on and initialized the UART, `uart_poweroff()` powers it off again. There is no use case for `uart_poweron()`.

maribu replied:
Oh, [it is actually documented](https://api.riot-os.org/group__drivers__periph__uart.html ):

> After initialization, the UART peripheral should be powered on and active. The UART can later be explicitly put to sleep and woken up by calling the uart_poweron() and uart_poweroff() functions. Once woken up using uart_poweron(), the UART should transparently continue it's previously configured operation.

Just not at the API of the functions. I still wonder why a low level driver is burdened with backing up the device state. Especially since in most case the compiler will not be able to optimize the RAM used for this out, even if it is never used.

benpicco commented:
Please squash!

benpicco commented:
keep squashing

maribu commented on cpu/rpx0xx/periph/uart.c:
```diff
@@ -67,7 +67,7 @@ void _set_symbolrate(uart_t uart, uint32_t baud)
     dev->UARTFBRD.reg = baud_fbrd;
 
-    io_reg_atomic_set(&(dev->UARTLCR_H.reg), 0);
+    dev->UARTLCR_H.reg = 0;
```
This shouldn't be needed, actually. (This also explains why this worked with the previous no-op atomic `|= 0` operation). Because:

1. `0x0` is the reset value of the register. So after the reset which is triggered during `uart_init()`, it will have this value anyway
2. In `uart_mode()` the register is overwritten anyway

I'll just drop this.

maribu commented:
I squashed again and added a fix for `uart_poweron()` / `uart_poweroff()`, the mixing pin access helper functions, as well as a corrected list of provided features .

I have to check with a logic analyzer to confirm that the `uart_poweron()` / `uart_poweroff()` is working and esoteric modes are correctly implemented.

Sadly, the UART is sending some garbage upon cold boot. (Warm boot it garbage-free.) And also a gargabe char is received upon boot (no matter if cold or warm boot). It would be nice to get rid of that, if possible.

maribu commented:
Tested the UART using `tests/periph_uart_mode` with the following patch applied:

```diff
diff --git a/tests/periph_uart_mode/Makefile b/tests/periph_uart_mode/Makefile
index 04882515c0..329a1ac637 100644
--- a/tests/periph_uart_mode/Makefile
+++ b/tests/periph_uart_mode/Makefile
@@ -2,8 +2,7 @@ include ../Makefile.tests_common
 
 FEATURES_REQUIRED += periph_uart
 FEATURES_REQUIRED += periph_uart_modecfg
-
-USEMODULE += xtimer
+FEATURES_OPTIONAL += periph_timer
 
 # Set this to prevent welcome message from printing and confusing output
 CFLAGS+=-DLOG_LEVEL=LOG_NONE
diff --git a/tests/periph_uart_mode/main.c b/tests/periph_uart_mode/main.c
index d15df67ba0..b37d44796b 100644
--- a/tests/periph_uart_mode/main.c
+++ b/tests/periph_uart_mode/main.c
@@ -22,7 +22,9 @@
 #include <string.h>
 #include <stdlib.h>
 
+#include "board.h"
 #include "periph/uart.h"
+#include "periph_conf.h"
 #include "stdio_uart.h"
 #include "xtimer.h"
 
@@ -52,6 +54,25 @@
 /* Stores each mode string for printing at the end of the test */
 static char mode_strings[TOTAL_OPTIONS][MODE_STR_LEN];
 
+static void _delay(void)
+{
+    if (IS_USED(MODULE_XTIMER)) {
+        xtimer_usleep(DELAY_US);
+    }
+    else {
+        /*
+         * As fallback for freshly ported boards with no timer drivers written
+         * yet, we just use the CPU to delay execution and assume that roughly
+         * 20 CPU cycles are spend per loop iteration.
+         *
+         * Note that the volatile qualifier disables compiler optimizations for
+         * all accesses to the counter variable. Without volatile, modern
+         * compilers would detect that the loop is only wasting CPU cycles and
+         * optimize it out - but here the wasting of CPU cycles is desired.
+         */
+        for (volatile uint32_t i = 0; i < CLOCK_CORECLOCK / 20; i++) { }
+    }
+}
 
 static void _get_mode(const uart_data_bits_t data_bits,
                       const uart_parity_t parity, const uart_stop_bits_t stop_bits, char* mode_str) {
@@ -153,7 +174,7 @@ int main(void)
                 if (status == UART_OK) {
                     results[ridx] = true;
                     printf("%s\n", mode_strings[ridx]);
-                    xtimer_usleep(DELAY_US);
+                    _delay();
                 }
                 else {
                     results[ridx] = false;

```

##### 5 Data Bits, No Parity, 1 Stop Bit

![Logic Analyzer Output](https://github.com/maribu/images/raw/master/RP2040-UART/5N1.png)


##### 5 Data Bits, No Parity, 2 Stop Bit

![Logic Analyzer Output](https://github.com/maribu/images/raw/master/RP2040-UART/5N2.png)

##### 5 Data Bits, Even Parity, 1 Stop Bit

![Logic Analyzer Output](https://github.com/maribu/images/raw/master/RP2040-UART/5E1.png)

##### 5 Data Bits, Even Parity, 2 Stop Bit

![Logic Analyzer Output](https://github.com/maribu/images/raw/master/RP2040-UART/5E2.png)

##### 6 Data Bits, No Parity, 1 Stop Bit

![Logic Analyzer Output](https://github.com/maribu/images/raw/master/RP2040-UART/6N1.png)

##### 7 Data Bits, No Parity, 1 Stop Bit

![Logic Analyzer Output](https://github.com/maribu/images/raw/master/RP2040-UART/7N1.png)

##### 8 Data Bits, No Parity, 1 Stop Bit

![Logic Analyzer Output](https://github.com/maribu/images/raw/master/RP2040-UART/8N1.png)

##### 8 Data Bits, No Parity, 2 Stop Bit

![Logic Analyzer Output](https://github.com/maribu/images/raw/master/RP2040-UART/8N2.png)

##### 8 Data Bits, Even Parity, 1 Stop Bit

![Logic Analyzer Output](https://github.com/maribu/images/raw/master/RP2040-UART/8E1.png)

##### 8 Data Bits, Even Parity, 2 Stop Bit

![Logic Analyzer Output](https://github.com/maribu/images/raw/master/RP2040-UART/8E2.png)

##### 8 Data Bits, Odd Parity, 1 Stop Bit

![Logic Analyzer Output](https://github.com/maribu/images/raw/master/RP2040-UART/8O1.png)

##### 8 Data Bits, Odd Parity, 2 Stop Bit

![Logic Analyzer Output](https://github.com/maribu/images/raw/master/RP2040-UART/8O2.png)

maribu commented:
In addition to all modes now working correctly (without sending one char with the previous config) the single garbage char previously send / transmitted on (cold) boot is now also gone.

maribu commented:
Output of the `tests/periph_gpio` when using https://github.com/RIOT-OS/RIOT/pull/16627 and the following patch applied:

```diff
diff --git a/tests/periph_gpio/main.c b/tests/periph_gpio/main.c
index 8011f45dc1..51e8469419 100644
--- a/tests/periph_gpio/main.c
+++ b/tests/periph_gpio/main.c
@@ -167,6 +167,7 @@ static int enable_int(int argc, char **argv)
     }
 
     po = atoi(argv[1]);
+    (void)po;
     pi = atoi(argv[2]);
 
     status = atoi(argv[3]);
```

```
> auto_test 0 2 0 3
2021-07-13 17:51:19,245 # auto_test 0 2 0 3
2021-07-13 17:51:19,246 # [START]
2021-07-13 17:51:19,253 # [SUCCESS]
```

benpicco commented on cpu/rpx0xx/periph/uart.c:
```diff
@@ -0,0 +1,280 @@
+    assert((unsigned)uart < UART_NUMOF);
+    _poweron(uart);
+    UART0_Type *dev = uart_config[uart].dev;
+    /* restore configuration registers */
+    dev->UARTIBRD.reg = uartibrd;
+    dev->UARTFBRD.reg = uartfbrd;
+    dev->UARTLCR_H.reg = uartlcr_h;
+    dev->UARTCR.reg = uartcr;
```
I think this is the only driver that properly implements this API :smile:

benpicco reviewed (approved):
Looks pretty good!
Please squash.

maribu commented:
Looks like Murdock tripped over another bug in `make info-boards-supported`, as there is no `periph_flashpage` feature provided by the pico. Locally, I cannot reproduce the issue :-(

benpicco commented:
Adding the missing `ROM_START_ADDR` and `CPU_FLASH_BASE` defines shouldn't hurt though

benpicco commented on cpu/rpx0xx/include/cpu_conf.h:
```diff
@@ -0,0 +1,47 @@
+ * @{
+ */
+#define CPU_DEFAULT_IRQ_PRIO            (1U)
+#define CPU_FLASH_BASE                  ROM_START_ADDR
```
ROM_START_ADDR is only defined as a `make` variable, not a C define.

maribu replied:
Did so via `CFLAGS += -D` to avoid having to sync to numbers, in case something changes.

maribu commented:
All green :-)

maribu commented:
(And I think the failing test previously might just have been the unit tests for `periph_flashpage`, which do not rely on the feature. So no bug in `make info-boards-supported`.)

maribu commented:
Thanks for the review! And thanks for the automatic GPIO test. It was very helpful in detecting a bug that would likely have gone unnoticed for quite some time!
